src/shared/Core/GenericHostProvider.cs

```diff
@@ -139,33 +139,19 @@ private async Task<ICredential> GetOAuthAccessToken(Uri remoteUri, string userNa
                 config.ClientSecret,
                 config.UseAuthHeader);
 
-            //
-            // Prepend "refresh_token" to the hostname to get a (hopefully) unique service name that
-            // doesn't clash with an existing credential service.
-            //
-            // Appending "/refresh_token" to the end of the remote URI may not always result in a unique
-            // service because users may set credential.useHttpPath and include "/refresh_token" as a
-            // path name.
-            //
-            string refreshService = new UriBuilder(remoteUri) { Host = $"refresh_token.{remoteUri.Host}" }
-                .Uri.AbsoluteUri.TrimEnd('/');
-
-            // Try to use a refresh token if we have one
-            ICredential refreshToken = Context.CredentialStore.Get(refreshService, userName);
+            if (!Context.CredentialStore.CanStoreOAuthRefreshToken) {
+                var refreshService = GetRefreshTokenServiceName(remoteUri);
+                refreshToken ??= Context.CredentialStore.Get(refreshService, userName)?.Password;
+            }
             if (refreshToken != null)
             {
+                Context.Trace.WriteLine("Refreshing OAuth token");
                 try
                 {
-                    var refreshResult = await client.GetTokenByRefreshTokenAsync(refreshToken.Password, CancellationToken.None);
-
-                    // Store new refresh token if we have been given one
-                    if (!string.IsNullOrWhiteSpace(refreshResult.RefreshToken))
-                    {
-                        Context.CredentialStore.AddOrUpdate(refreshService, refreshToken.Account, refreshToken.Password);
-                    }
+                    var refreshResult = await client.GetTokenByRefreshTokenAsync(refreshToken, CancellationToken.None);
 
                     // Return the new access token
-                    return new GitCredential(oauthUser,refreshResult.AccessToken);
+                    return new GitCredential(refreshResult, oauthUser);

```

Comment:
Delaying the save of a `refresh_token` will likely result in issues.

New value must be saved, even if the generated `auth_token` may be invalid for the requested operation.
This means we'd have to check (and rely on git) to also provide the _new_ `refresh_token` and store its value during an `erase` callback (needs ugly hack in `EraseCredentialAsync`).

Granted the current state (without #1838) is _also_ broken.
But I think this is still the correct location to issue a store.
Otherwise we risk client desync due to missed `refresh_token` value updates.